tools/wake-unit/unit.h

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+#include <vector>
+#include <csetjmp>
+#include "util/colour.h"
+
+// Represents an error message to display to the user
+struct ErrorMessage {
+  // Test and failure location information
+  const char* test_name;
+  const char* file;
+  int line;
+
+  // The generated error message with precise details
+  std::stringstream predicate_error;
+
+  // The error message supplied by the user
+  std::stringstream user_error;
+};
+
+// A class that handles the return value from EXPECT_* and ASSERT_*
+// On assert failure it longjumps back to the test harness. Allows
+// the user to add specilized messages to errors like gtest. If the
+// the test did not fail, the user supplied messages are ignored.
+struct TestStream {
+  std::stringstream* ss;
+  std::jmp_buf* assert_throw;
+  TestStream(std::stringstream* ss_, std::jmp_buf* assert_) : ss(ss_), assert_throw(assert_) {}
+  ~TestStream() {
+    if (assert_throw) std::longjmp(*assert_throw, 1);
+  }
+  template <class T>
+  TestStream& operator<<(T&& x) {
+    if (ss) *ss << x;
+    return *this;
+  }
+};
+
+// Public:
+struct TestLogger {
+  std::vector<ErrorMessage> errors;
+  std::jmp_buf return_jmp_buffer;
+  const char* test_name = nullptr;
+
+private:
+  TestStream fail(ErrorMessage& err, bool assert) {
+    return TestStream(&err.user_error, assert ? &return_jmp_buffer : nullptr);
+  }
+
+  // TODO: Make this work with unicode
+  static std::string quote(std::string str) {
```

Comment:
Would `json_escape` have worked for this?

Reply:
Yeah I think I'll do that instead, I dislike this function as is.